core/py/src/braintrust_core/json.py

```diff
@@ -0,0 +1,69 @@
+import dataclasses
+import json
+from typing import Any, Dict, Optional
+
+
+class JSONSerializationError(Exception):
+    pass
+
+
+def check_json_serializable(obj: Any) -> None:
+    try:
+        _ = json.dumps(obj)
+    except Exception as e:
+        raise JSONSerializationError from e
+
+
+def is_json_serializable(obj: Any) -> bool:
+    try:
+        check_json_serializable(obj)
+        return True
+    except JSONSerializationError:
+        return False
+
+
+def dump_object(obj: Any) -> Any:
+    """Convert an object to a JSON-serializable object if it is an instance of `dataclasses.dataclass` or `pydantic.BaseModel`."""
+    if dataclasses.is_dataclass(obj) and not isinstance(obj, type):
+        return dataclasses.asdict(obj)
+
+    # Attempt to dump a Pydantic v2 `BaseModel`.
+    try:
+        return obj.model_dump()
+    except (AttributeError, TypeError):
+        pass
+
+    # Attempt to dump a Pydantic v1 `BaseModel`.
+    try:
+        return obj.dict()
+    except (AttributeError, TypeError):
+        pass
+
+    return obj
+
+
+def make_json_serializable(obj: Any) -> Any:
```

Comment:
I'm a little worried about the efficiency of this function, because it's trying to JSON serialize the object at the very start to check if it's JSON serializable, and then again at the very end. We should be careful (and hopefully very clear) about how many times we try JSON-serializing the objects they give us.

I wonder if an easier way to do all of this is to utilize the `default` method of the json encoding library: https://docs.python.org/3/library/json.html#json.JSONEncoder.

Maybe we could create our own JSON encoder as
```
class BraintrustJSONEncoder(json.JSONEncoder):
    def default(self, o):
        # Your dump_obj implementation
        ...
        # In case none of the cases match
        return super().default(o)
```

And use that everywhere instead of the standard one?